enclave-runtime/src/ocall/sidechain_ocall.rs

```diff
@@ -65,4 +67,40 @@ impl EnclaveSidechainOCallApi for OcallApi {
 
 		Ok(())
 	}
+
+	fn fetch_sidechain_blocks_from_peer<SignedSidechainBlock: Decode>(
+		&self,
+		last_known_block_hash: BlockHash,
+		shard_identifier: ShardIdentifier,
+	) -> SgxResult<Vec<SignedSidechainBlock>> {
+		let mut rt: sgx_status_t = sgx_status_t::SGX_ERROR_UNEXPECTED;
+		let last_known_block_hash_encoded = last_known_block_hash.encode();
+		let shard_identifier_encoded = shard_identifier.encode();
+
+		// We have to pre-allocate the vector and hope it's large enough
+		let mut signed_blocks_encoded: Vec<u8> = vec![0; 4096 * 16];
```

Comment:
This is what worries me a bit: In order to get the sidechain blocks over the o-call interface, we have to pre-allocate the vector into which the block data is filled. I think there's no other elegant solution to this? An alternative would be to have an additional o-call to query the size first, so we can then allocate exactly the right amount. But it would basically double the number of calls into our block storage.

Reply:
Ouf, you're right. That's quite ugly. I do see a different solution, but that might require some redesigning:

Instead of fetching the blocks directly with this ocall, you could just and kick off a fetcher function on the untrusted side which is retrieves the sidechain blocks and buffers them on the untrusted side  (not sure if buffering is needed, there might be a simpler solution). For each retrieved sidechain block, this untrusted function would do an ecall to import the block (I guess that could be the same ecall as for the gossiped sidechain blocks).

Reply:
Hmm I think that would be more complex to do, because we would go from a pull model to a push model (from the enclave perspective), which is not easy to do in the current PeerBlockSyncer.
An alternative would be to extend the o-call to have an additional output parameter, which is the total buffer size required. And this will be set in case the default one is not enough. We can then detect this in our o-call and adjust our buffer size accordingly and make the call again.

Reply:
I do agree that this is more complex. But simply adjusting the buffer to an arbitrary amount I also don't think is sensible - except if we limit our sidechain buffering to a reasonable size which we can import in one go. 
Either way - I think this issue is worth it's own PR and discussion, so I suggest to open a new issue.

Reply:
Issue created -> https://github.com/integritee-network/worker/issues/621